### PR TITLE
feat(eks): spot support for managed nodegroups

### LINKED ISSUE #11827: (aws-eks): managed nodegroup for spot instances
Amazon EKS now supports provisioning and managing EC2 Spot Instances in managed node groups
https://aws.amazon.com/tw/blogs/containers/amazon-eks-now-supports-provisioning-and-managing-ec2-spot-instances-in-managed-node-groups/

`eksctl` just introduced the new `--spot` flag which seems to create a new launch template with spot options and pass this template to create the spot-only nodegroup.

I was wondering how to create similar experience with `aws-eks`.  The [Nodegroup](https://github.com/aws/aws-cdk/blob/869c884a9bd02a5fa116a8339ef7a6cedbeb33ac/packages/%40aws-cdk/aws-eks/lib/managed-nodegroup.ts#L201) L2 construct actually can accept the launch template as the construct property, however, users need to bake their own LT with spot options before they can pass it to `NodegroupProps`.  It doesn't make sense to add a new `spot` property for `NodegroupProps` as this is actually a high level abstraction, but making people easily create spot nodegroup with CDK is really helpful.

Some options:

1. create  `aws-eks-patterns` L3 and make `SpotNodegroup` a L3 construct
2. let's create a 3rd party construct lib like `cdk-eks-spot-nodegroup`

I'd vote option 1.

@iliapolo wdyt?

ref: https://docs.aws.amazon.com/eks/latest/userguide/managed-node-groups.html

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by pahud**
This PR adds the `CapacityType` support and allows users to create Spot managed node groups for Amazon EKS.

1. The `CapacityType` attribute is supported by cloudformation but not yet documented. We tentatively use addPropertyOverride() to enable it.
2. `instanceType` will be deprecated and we introduced the new `instanceTypes`
3. `instanceTypes` with different CPU architectures will throw an error.
4. `amiType` is still optional, however, when specified, incorrect `amiType` will throw the error.
5. According to the [document](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html#cfn-eks-nodegroup-instancetypes), we are allowed to specify instance type(s) in either `instanceTypes` property or launch template but not both. As we can't check the content of the launch template passed in, we allow `instanceTypes` and launch template both specified and encourage to use `instanceTypes` when possible.

## Sample

```ts
cluster.addNodegroupCapacity('extra-ng-spot', {
  instanceTypes: [
    new ec2.InstanceType('c5.large'),
    new ec2.InstanceType('c5a.large'),
    new ec2.InstanceType('c5d.large'),
  ],
  minSize: 3,
  capacityType: eks.CapacityType.SPOT,
});
```

Closes #11827 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by pahud** [comment]
@iliapolo I'm ready for the first round.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 1]
```diff
@@ -271,11 +291,11 @@ export class Nodegroup extends Resource implements INodegroup {
       nodegroupName: props.nodegroupName,
       nodeRole: this.role.roleArn,
       subnets: this.cluster.vpc.selectSubnets(props.subnets).subnetIds,
-      amiType: props.amiType ?? (props.instanceType ? getAmiTypeForInstanceType(props.instanceType).toString() :
+      amiType: props.amiType ?? (props.instanceType ? getAmiTypeForInstanceType(props.instanceType[0]).toString() :
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:294]: What if the array contains different types? is choosing `0` the right thing to do?

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 2]
```diff
@@ -139,7 +153,7 @@ export interface NodegroupOptions {
    *
    * @default t3.medium
    */
-  readonly instanceType?: InstanceType;
+  readonly instanceType?: InstanceType[];
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:156]: Let's not break the api. Add another property called `additionalInstanceTypes` and mention that when `SPOT` is used, it is also recommended to provide additional types using this property.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 3]
```diff
@@ -291,17 +311,21 @@ export class Nodegroup extends Resource implements INodegroup {
...
+      resource.addPropertyOverride('CapacityType', props.capacityType.valueOf());
+    }
+
     if (props.launchTemplateSpec) {
       if (props.diskSize) {
         // see - https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html
         // and https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html#cfn-eks-nodegroup-disksize
         throw new Error('diskSize must be specified within the launch template');
       }
-      if (props.instanceType) {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:300]: Why did you remove this validation?

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.expected.json [hunk 4]
```diff
@@ -3208,18 +3215,117 @@
             "Ref": "VpcPrivateSubnet3SubnetF258B56E"
           }
         ],
-        "AmiType": "AL2_x86_64",
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/integ.eks-cluster.expected.json:3211]: Is this expected..?

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 3 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:300]: According to this:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html#cfn-eks-nodegroup-instancetypes

>If you specify launchTemplate, then you can specify zero or one instance type in your launch template or you can specify 0-20 instance types for instanceTypes. If however, you specify an instance type in your launch template and specify any instanceTypes, the node group deployment will fail. If you don't specify an instance type in a launch template or for instanceTypes, then t3.medium is used, by default. If you specify Spot for capacityType, then we recommend specifying multiple values for instanceTypes.

If we design the `Nodegroup` API like

```ts
new Nodegroup(..., {
   instanceType: ec2.InstanceType,
   additionalInstanceTypes: ec2.InstanceType[],
   launchTemplateSpec,
});
```

In this case, one or multipe instance types may also be defined in  the`launchTemplate` and this will cause failure according to the document. However, This should also be accepted as user may leverage some other  features from LT while still defining the multiple instance types in `additionalInstanceTypes`.

As we can't check the content of the LT passed in and we encourage users to define instance types in `additionalInstanceTypes`, we probably should skip the check. Ideas?

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 2 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:156]: OK. So if we design the API like:

```ts
readonly instanceType: InstanceType;
readonly additonalInstanceTypes?: InstanceType[];
```

To create MNG with `t3.large`, `c5.large`, `m5.large` :

```ts
new Nodegroup(stack, id, {
   instanceType: new ec2.InstanceType('t3.large'),
   additionalInstanceTypes: [
      new ec2.InstanceType('c5.large'),
      new ec2.InstanceType('m5.large'),
  ],
  capacityType: eks.CapacityType.SPOT,
});
```

And we still determine the `AmiType` from `instanceType`, right?

**Comment by pahud** [comment]
Hi @iliapolo 

Please check my comments before I continue revise it. Thanks.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 2 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:156]: Well i'm not sure, what's going to happen if the AMI's are conflicting? I imagine this shouldn't happen often since the distinction is between graviton/gpu/inference instances.

What if we validate the all instance types resolve to the same AMI? and throw if they result in a conflict.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 3 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:300]: Ok so basically a user can specify a launch template without any instance types, and only specify them in the `InstanceTypes` property of the nodegroup, in which the deployment will succeed right? If thats the case then yes we should remove this validation 👍

**Comment by iliapolo** [comment]
@pahud see my responses

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 2 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:156]: Also i'm thinking we can probably go with `instanceTypes` (instead of `additionalInstanceTypes`), and deprecate `instanceType`. This would be clearer I think, and will map correctly to the underlying CFN properties.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 1 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: Addressed. Now we are using a function to determine the `amiType`.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 2 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: Yes. `instanceTypes` now.

**Comment by pahud** [comment]
@iliapolo  I am ready for the 2nd round.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 5]
```diff
@@ -246,13 +273,24 @@ export class Nodegroup extends Resource implements INodegroup {
     this.maxSize = props.maxSize ?? this.desiredSize;
     this.minSize = props.minSize ?? 1;
 
+    const DEFAULT_INSTANCE_TYPE = new InstanceType('t3.medium');
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:276]: I think we should make it:
```typescript
public static readonly DEFAULT_INSTANCE_TYPE = new InstanceType('t3.medium');
```

Or we should make it an array as the prop is actually an array `instanceTypes`
```typescript
public static readonly DEFAULT_INSTANCE_TYPES = [new InstanceType('t3.medium')];
```

ref: https://github.com/aws/aws-cdk/blob/8f6f9a8678496e131a43ca4c76e561d50a0a0de8/packages/%40aws-cdk/aws-docdb/lib/cluster.ts#L180

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 6]
```diff
@@ -135,11 +149,18 @@ export interface NodegroupOptions {
   /**
    * The instance type to use for your node group. Currently, you can specify a single instance type for a node group.
    * The default value for this parameter is `t3.medium`. If you choose a GPU instance type, be sure to specify the
-   * `AL2_x86_64_GPU` with the amiType parameter.
+   * `AL2_x86_64_GPU` with the amiType parameter. This property will be ignored if `instanceTypes` is defined.
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:152]: It's better to throw an error if both `instanceType` and `instanceTypes` is used. Will make the deprecation easier since we won't have to maintain this dual scenario.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 7]
```diff
@@ -340,3 +385,12 @@ function getAmiTypeForInstanceType(instanceType: InstanceType) {
           NodegroupAmiType.AL2_X86_64;
 }
 
+function determineAmiTypes(instanceType: InstanceType[]) {
+  const amiTypes = instanceType.map(i =>getAmiTypeForInstanceType(i));
+  const uniq = [...new Set(amiTypes)];
+  if (uniq.length > 1) {
+    throw new Error('instanceTypes of different CPU architectures not allowed');
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:392]: So is this effectively true from an EKS perspective as well? Making sure we are not introducing some artificial limitation now. 

Also - make this function return the set, and throw the necessary errors in the main control flow.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 8]
```diff
@@ -253,6 +284,15 @@ export class Nodegroup extends Resource implements INodegroup {
       throw new Error(`Minimum capacity ${this.minSize} can't be greater than desired size ${this.desiredSize}`);
     }
 
+    if (props.instanceType) {
+      Annotations.of(this).addWarning('instanceType will be deprecated, remember to use instanceTypes instead.');
+    }
+    const instanceTypes = props.instanceTypes ?? (props.instanceType ? [props.instanceType] : Nodegroup.DEFAULT_INSTANCE_TYPES);
+    const determinedAmiType = determineAmiTypes(instanceTypes);
+    if (props.amiType && props.amiType !== determinedAmiType) {
+      throw new Error(`amiType is not correct - should be ${determinedAmiType}`);
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:293]:
```suggestion
      throw new Error(`The specified AMI does not match the instance types architecture, either specify ${determinedAmiType} or dont specify any`);
```

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 9]
```diff
@@ -291,17 +332,21 @@ export class Nodegroup extends Resource implements INodegroup {
       tags: props.tags,
     });
 
+    if (props.capacityType) {
+      resource.addPropertyOverride('CapacityType', props.capacityType.valueOf());
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:336]: Lets maybe already create an issue for this so we don't forget? i'd really want to get rid of this once we can.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 7 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:392]: Yes, I've tried this with @guan840912 from CFN's perspective and it returns failure. So we have to throw an error here.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 9 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:346]: Yes. Noted https://github.com/aws/aws-cdk/issues/12154

**Comment by pahud** [comment]
Hi @iliapolo 

All addressed. Please check it out again.

File: packages/@aws-cdk/aws-eks/test/test.nodegroup.ts [hunk 10]
```diff
@@ -206,6 +227,185 @@ export = {
...
+        'm5.large',
+        't3.large',
+        'c5.large',
+      ],
+      CapacityType: 'SPOT',
+    },
+    ));
+    test.done();
+  },
+  'create nodegroup with ondemand capacity type'(test: Test) {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/test.nodegroup.ts:289]: Isn't this a duplicate of `create nodegroup with on-demand capacity type` on line 230?

File: packages/@aws-cdk/aws-eks/test/test.nodegroup.ts [hunk 11]
```diff
@@ -206,6 +227,185 @@ export = {
...
+      instanceTypes: [
+        new ec2.InstanceType('m5.large'),
+        new ec2.InstanceType('t3.large'),
+        new ec2.InstanceType('c5.large'),
+      ],
+      capacityType: eks.CapacityType.SPOT,
+    }), /"instanceType is deprecated, please use "instanceTypes" only/);
+    test.done();
+  },
+  'create nodegroup with either instanceTypes or instanceType defined'(test: Test) {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/test.nodegroup.ts:344]:
```suggestion
  'create nodegroup with neither instanceTypes nor instanceType defined'(test: Test) {
```

You mean?

File: packages/@aws-cdk/aws-eks/test/test.nodegroup.ts (hunk 10 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/test/test.nodegroup.ts]: @iliapolo 

Let me change the name to `create nodegroup with on-demand capacity type and multiple instances`

so we have two test cases:

1.  on-demand capacity type with one instance in `instanceTypes`
2. on-demand capacity type with multiple instances in `instanceTypes`, which is also accepted by CFN. Under the hood the CFN will create an ASG with launch template with multiple instance types but only the first one will be selected. It's not documented specificly in CFN but it's also accepted.

**Comment by pahud** [comment]
Hi @iliapolo 

All fixed. Please check it out again.

File: packages/@aws-cdk/aws-eks/test/test.nodegroup.ts (hunk 10 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/test/test.nodegroup.ts]: @pahud Wait, creating an `on-demand` node group with multiple instance types will result in a launch template with just the first instance type being selected? ignoring all the others?

Comment (pahud) [reply packages/@aws-cdk/aws-eks/test/test.nodegroup.ts]: Hold on a few minutes. Let me paste some screenshots here later.

Comment (pahud) [reply packages/@aws-cdk/aws-eks/test/test.nodegroup.ts]: If we created the MNG like this


```ts
this.cluster.addNodegroupCapacity('MNG', {
  capacityType: eks.CapacityType.ON_DEMAND,
  desiredSize: 10,
  instanceTypes: [
    new ec2.InstanceType('t3.large'),
    new ec2.InstanceType('m5.large'),
    new ec2.InstanceType('c5.large'),
  ],
})
```

We actually synthesizes to this

```yaml
  ClusterNodegroupMNGF9B82D86:
    Type: AWS::EKS::Nodegroup
    Properties:
      ClusterName:
        Ref: Cluster9EE0221C
      NodeRole:
        Fn::GetAtt:
          - ClusterNodegroupMNGNodeGroupRole3AB840AB
          - Arn
      Subnets:
        - subnet-05b946a306d81274b
        - subnet-0ae2846156ce0118c
        - subnet-063ca7ed4aacdbad7
      AmiType: AL2_x86_64
      ForceUpdateEnabled: true
      InstanceTypes:
        - t3.large
        - m5.large
        - c5.large
      ScalingConfig:
        DesiredSize: 10
        MaxSize: 10
        MinSize: 1
      CapacityType: ON_DEMAND
```

And the MNG will create an ASG with LT with 100% OD + 0% Spot like:

圖片

圖片

圖片

And all the instances will be the first instance type

圖片


I believe this is expected from ASG's perspective when ASG+LT for on-demand instance types, ASG will not span multiple instance types.

Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/test/test.nodegroup.ts]: @pahud Its not that all the other instance types are ignored, the distribution of instance types across the ASG depends on the allocation strategy, which for on-demand is [`prioritized `](https://docs.aws.amazon.com/autoscaling/ec2/userguide/asg-purchase-options.html#asg-allocation-strategies).

To my understanding this means it will try to fulfill all the capacity by trying the first instance type, but **will**  use other types if there is still a capacity shortage and the first type is not depleted.

I kind of lost sight of why we got into this discussion honestly :) - does this make a difference to us somehow?

**Comment by iliapolo** [comment]
@pahud I'm good with merging i think once the conflicts are resolved.

**Comment by pahud** [comment]
Hi @iliapolo 

All fixed now. :-)

File: packages/@aws-cdk/aws-eks/README.md [hunk 12]
```diff
@@ -201,14 +201,35 @@ const cluster = new eks.Cluster(this, 'HelloEKS', {
...
   minSize: 4,
   diskSize: 100,
   amiType: eks.NodegroupAmiType.AL2_X86_64_GPU,
   ...
 });
 ```
 
+#### Spot Instances Support
+
+Use `capacityType` to create the Spot managed node groups. To maximize the availability of your applications while using
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/README.md:214]:
```suggestion
Use `capacityType` to create managed node groups comprised of spot instances. To maximize the availability of your applications while using
```

File: packages/@aws-cdk/aws-eks/README.md [hunk 13]
```diff
@@ -201,14 +201,35 @@ const cluster = new eks.Cluster(this, 'HelloEKS', {
...
+Spot Instances, we recommend that you configure a Spot managed node group to use multiple instance types with `instanceTypes`.
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/README.md:215]:
```suggestion
Spot Instances, we recommend that you configure a Spot managed node group to use multiple instance types with the `instanceTypes` property. 
```

File: packages/@aws-cdk/aws-eks/README.md [hunk 14]
```diff
@@ -236,7 +257,9 @@ cluster.addNodegroupCapacity('extra-ng', {
 });
 ```
 
-> For more details visit [Launch Template Support](https://docs.aws.amazon.com/en_ca/eks/latest/userguide/launch-templates.html).
+You may specify one or multiple instance types in either `instanceTypes` property of `NodeGroup` or in the launch template but not both.
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/README.md:260]:
```suggestion
You may specify one or instance types in either the `instanceTypes` property of `NodeGroup` or in the launch template, **but not both**.
```

**Review by iliapolo — APPROVED** [review]
